src/NodeAnalyzer/RelationshipAnalyzer.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace RectorLaravel\NodeAnalyzer;
+
+use InvalidArgumentException;
+use PHPStan\Analyser\Scope;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\Type;
+
+class RelationshipAnalyzer
+{
+    protected static function relationType(): ObjectType
+    {
+        return new ObjectType('Illuminate\Database\Eloquent\Relations\Relation');
+    }
+
+    protected static function modelType(): ObjectType
+    {
+        return new ObjectType('Illuminate\Database\Eloquent\Model');
+    }
+
+    /**
+     * Resolve the Related Model of the Relationship.
+     *
+     * @return ObjectType|null
+     */
+    public function resolveRelatedForRelation(Type $objectType, Scope $scope): ?Type
+    {
+        if ($objectType->isObject()->no() || $objectType->isSuperTypeOf(self::relationType())->no()) {
+            throw new InvalidArgumentException('Object type must be an Eloquent relation.');
+        }
+
+        $extendedPropertyReflection = $objectType->getInstanceProperty('related', $scope);
+        $modelType = $extendedPropertyReflection->getReadableType();
+
+        if ($modelType->isSuperTypeOf(self::modelType())->no()) {
+            return null;
+        }
+
+        /** @phpstan-ignore return.type */
+        return $modelType;
+    }
+
+    /**
+     * Resolve the Parent Model of the Relationship.
+     *
+     * @return ObjectType|null
+     */
+    public function resolveParentForRelation(Type $objectType, Scope $scope): ?Type
+    {
+        if ($objectType->isObject()->no() || $objectType->isSuperTypeOf(self::relationType())->no()) {
+            throw new InvalidArgumentException('Object type must be an Eloquent relation.');
+        }
+
+        $extendedPropertyReflection = $objectType->getInstanceProperty('parent', $scope);
+        $modelType = $extendedPropertyReflection->getReadableType();
+
+        if ($modelType->isSuperTypeOf(self::modelType())->no()) {
+            return null;
+        }
+
+        /** @phpstan-ignore return.type */
+        return $modelType;
+    }
```

Comment:
I doubt this will ever be needed (I've never needed the parent in Larastan), but here's the correct version


```suggestion
    /**
     * Resolve the Parent Model of the Relationship.
     *
     * @return ObjectType|null
     */
    public function resolveParentForRelation(Type $objectType): ?Type
    {
        $modelType = $objectType->getTemplateType(Relation::class, 'TDeclaringModel');
        
        if (self::modelType()->isSuperTypeOf($modelType)->no()) {
            return null;
        }
        
        /** @phpstan-ignore return.type */
        return $modelType;
    }
```